# Analysis scripts for Baez-Ortega et al., 2019
# Step 11. Selection analyses

# Adrian Baez-Ortega, 2019


# TO RUN THIS SCRIPT IN THE TERMINAL
# ----------------------------------
# Run the commands below in the terminal, replacing '/path/to/TCG2019' with
# the path to the TCG2019 directory.
#
#    cd /path/to/TCG2019
#    Rscript scripts/11_SelectionAnalyses.R


# TO RUN THIS SCRIPT IN RSTUDIO
# -----------------------------
# Before starting, run the line below in RStudio, replacing '/path/to/TCG2019' with
# the path to the TCG2019 directory.
#
#    setwd("path/to/TCG2019")


# If the paths to the input or output files differ from the ones
# used in this script, they may be updated by modifying the lines below.

INPUT = list(
    
    # Path to list of essential genes
    ESS.GENES = file.path("data", "original", "Essential_Genes.tsv"),
    
    # Path to file of GO terms per gene
    CF3.GO = file.path("data", "original", "BioMart_CanFam3.1_GOs.RData"),
    
    # Path to CanFam3.1 RefCDS database for dNdScv
    CF3.CDS = file.path("data", "original", "dNdScv_RefCDS_Dog_CanFam3.1.RData"),
    
    # Path to CanFam3.1 covariates for dNdScv (translated via homology from hg19)
    CF3.COV = file.path("data", "original", "dNdScv_Covariates_Dog_CanFam3.1.RData"),
    
    # Path to recurrent mutation tables from mutree
    REC.MUT = file.path("data", "original", "Mutree_Recurrent_Mutations.RData"),
    
    # Path to input file generated by script 2_ImportVariants.R
    VAR.TABLES = file.path("data", "processed", "Variant_Tables.RData"),
    
    # Path to input file generated by script 3_ProcessAnnotation.R
    VAR.ANNOT = file.path("data", "processed", "Variant_Annotation.RData"),
    
    # Path to input file generated by script 8_DinucAnalyses.R
    DINUCS = file.path("data", "processed", "Dinucleotides.RData"),
    
    # Path to input file generated by script 10_ExpressionAnalyses.R
    TX.ABUND = file.path("data", "processed", "Mean_Transcript_Abundances.RData")
    
)

OUTPUT = list(
    
    # Path to output PDF
    PLOTS = file.path("output", "dNdS_Global_Plot.pdf"),
    
    # Path to output tables of global dN/dS
    GLOBAL.TBL = file.path("output", "dNdS_Global_Tables.tsv"),
    
    # Path to output table of genewise dN/dS
    GENE.TBL = file.path("output", "dNdS_Genewise_Table.tsv"),
    
    # Path to output RData file
    RDATA = file.path("data", "processed", "Selection_Results.RData")
    
)


# Print input and output file paths
cat("\nInput files:\n")
for (name in INPUT) {
    cat(paste0("   ", name), sep="\n")
}
cat("\nOutput files:")
for (name in OUTPUT) {
    cat("\n  ", name)
}
cat("\n\n")


# Function for suppressing terminal output
quiet = function(x) {
    f = file(tempfile(), open="wt")
    sink(f) 
    sink(f, type="message")
    on.exit(sink(type="message"))
    on.exit(sink(), add=TRUE) 
    suppressWarnings(invisible(force(x)))
} 


# Load packages
PACKAGES = c("dndscv")
cat("Loading packages:", paste(PACKAGES, collapse=", "), "\n")
for (package in PACKAGES) {
    quiet(library(package, character.only=T))
}


# Load input data
cat("Loading data...\n")
load(INPUT$CF3.GO)
load(INPUT$CF3.CDS)
load(INPUT$CF3.COV)
load(INPUT$REC.MUT)
load(INPUT$VAR.TABLES)
load(INPUT$VAR.ANNOT)
load(INPUT$DINUCS)
load(INPUT$TX.ABUND)
essential.genes = read.table(INPUT$ESS.GENES, sep="\t", header=TRUE, stringsAsFactors=FALSE)


## (1) Build mutations table
cat("\nBuilding table of mutations...\n")

# Duplicate manually validated recurrent mutations
recurrent.metaidx = NULL
for (i in 1:nrow(identical.valid.table)) {
    # Get original mutree mutation
    # Nonsynonymous case
    idx = identical.valid.table$Mut_string[i] == rec.mutations.filtered$String
    if (any(idx)) {
        stopifnot(sum(idx) > 1)
        muts = rec.mutations.filtered[idx,]
    } 
    # Synonymous case
    else {
        idx = identical.valid.table$Mut_string[i] == syn.ident.mutations.filt$String
        stopifnot(sum(idx) > 1)
        muts = syn.ident.mutations.filt[idx,]
    }
    # Repeat variant meta-index as many times as independent occurrences
    recurrent.metaidx = c(recurrent.metaidx,
                          rep(identical.valid.table$Meta_index[i], nrow(muts)))
}
tonly.plus.recurrent.idx = c(which(tumour.only.idx), recurrent.metaidx)
stopifnot(length(unique(tonly.plus.recurrent.idx)) == sum(tumour.only.idx))

mutations = data.frame("sampleID" = "CTVT",
                       "chr" = as.character(snvs.metadata$CHROM[tonly.plus.recurrent.idx]),
                       "pos" = snvs.metadata$POS[tonly.plus.recurrent.idx],
                       "ref" = as.character(snvs.metadata$REF[tonly.plus.recurrent.idx]),
                       "mut" = as.character(snvs.metadata$ALT[tonly.plus.recurrent.idx]),
                       stringsAsFactors=FALSE)

# Discard mutations corresponding to dinucleotide variants
# (only dinucleotides for which both mutations are found)
mut.pos = paste(mutations$chr, mutations$pos, sep=":")
mut.pos.prev = paste(mutations$chr, mutations$pos - 1, sep=":")
dnt.pos = paste(dnvs.metadata$CHROM[tumour.only.dnvs.idx],
                dnvs.metadata$POS[tumour.only.dnvs.idx], sep=":")

dnt.idx = mut.pos %in% dnt.pos | mut.pos.prev %in% dnt.pos
dnt.idx[dnt.idx] = c(diff(mutations$pos[dnt.idx]), 0) == 1 | c(0, diff(mutations$pos[dnt.idx])) == 1

# Discard mutations in previously significant genes that were found to be artefacts
# (by visual validation)
artefact.rows = c("242675", "242676", "242677", "242678", "242679", "242680", "242681", "242682",
                  "242683", "242685", "242686", "242688", "242689", "227629", "227630", "227631",
                  "227632", "244619", "244621", "244623", "244624", "244625", "244627", "244628",
                  "244629", "244630", "244631", "244632", "244633", "244634", "244637", "244638",
                  "244639", "244641", "244642", "244643", "244644")

artefact.idx = paste0(mutations$chr, ":", mutations$pos, mutations$mut) %in%
    paste0(annot.snvs.tonly[artefact.rows, "Chrom"], ":",
           annot.snvs.tonly[artefact.rows, "Pos"], annot.snvs.tonly[artefact.rows, "Allele"])

stopifnot(length(dnt.idx) == nrow(mutations) & length(artefact.idx) == nrow(mutations))
mutations = mutations[!dnt.idx & !artefact.idx, ]


## (2) Run dNdScv on all genes (with CanFam3.1 covariates)
cat("\nCalculating global and genewise dN/dS in all genes...\n")

dnds.all.cv = suppressWarnings(dndscv(mutations, refdb=INPUT$CF3.CDS, cv=cf3.covs,
                                      max_muts_per_gene_per_sample=Inf,
                                      max_coding_muts_per_sample=Inf, maxcovs=Inf))


## (3) Run dNdScv on essential genes
cat("\nCalculating global dN/dS in essential genes...\n")

# Create lists of essential genes (All, CN=1, CN>1)
gene.names.db = sapply(RefCDS, function(x) x$gene_name)
gene.ids.db = sapply(RefCDS, function(x) x$gene_id)
essential.genes.dndscv = NULL

essential.genes.dndscv$ALL = gene.names.db[match(essential.genes$Gene, gene.ids.db)]
essential.genes.dndscv$ALL = essential.genes.dndscv$ALL[!is.na(essential.genes.dndscv$ALL)]

idx = essential.genes$CN24T == 1 & essential.genes$CN79T == 1
essential.genes.dndscv$`CN=1` = gene.names.db[match(essential.genes$Gene[idx], gene.ids.db)]
essential.genes.dndscv$`CN=1` = essential.genes.dndscv$`CN=1`[!is.na(essential.genes.dndscv$`CN=1`)]

idx = essential.genes$CN24T >= 2 & essential.genes$CN79T == essential.genes$CN24T
essential.genes.dndscv$`CN>1` = gene.names.db[match(essential.genes$Gene[idx], gene.ids.db)]
essential.genes.dndscv$`CN>1` = essential.genes.dndscv$`CN>1`[!is.na(essential.genes.dndscv$`CN>1`)]

# Run dNdScv
dnds.essential.cv = lapply(1:length(essential.genes.dndscv), function(i) {
    cat("\nGroup: ", names(essential.genes.dndscv)[i], "\n", sep="")
    suppressWarnings(dndscv(mutations, cv=cf3.covs, refdb=INPUT$CF3.CDS,
                            max_muts_per_gene_per_sample=Inf,
                            max_coding_muts_per_sample=Inf, maxcovs=Inf,
                            gene_list=essential.genes.dndscv[[i]], outp=1))
})
names(dnds.essential.cv) = names(essential.genes.dndscv)


## (4) Run dNdScv per gene copy number group
cat("\nCalculating global dN/dS per gene copy number...\n")

# Calculate mean CN per gene
gene.cn = t(sapply(unique(annot.snvs.tonly$Gene), function(gene) {
    gene = as.character(gene)
    sym = annot.snvs.tonly$Symbol[match(gene, annot.snvs.tonly$Gene)]
    var.idx = annot.snvs.tonly$Gene == gene
    cn.24 = mean(annot.snvs.tonly$CN_24T[var.idx], na.rm=T)
    cn.79 = mean(annot.snvs.tonly$CN_79T[var.idx], na.rm=T)
    c("Gene"=gene, "Symbol"=sym, "Average 24T CN"=cn.24, "Average 79T CN"=cn.79)
}))

# Make gene lists
cn.ranges = list("CN=1"=c(0.5, 1.5), "CN=2"=c(1.5, 2.5), "CN>2"=c(2.5, Inf))
cn.groups.dndscv = lapply(cn.ranges, function(rng) {
    idx = gene.cn[, "Average 24T CN"] >= rng[1] &
        gene.cn[, "Average 24T CN"] < rng[2] &
        gene.cn[, "Average 79T CN"] >= rng[1] &
        gene.cn[, "Average 79T CN"] < rng[2]
    gns = gene.names.db[match(gene.cn[idx, "Gene"], gene.ids.db)]
    gns[!is.na(gns)]
})

# Add list of non-essential genes with CN=1
rng = cn.ranges$`CN=1`
idx = gene.cn[, "Average 24T CN"] >= rng[1] &
    gene.cn[, "Average 24T CN"] < rng[2] &
    gene.cn[, "Average 79T CN"] >= rng[1] &
    gene.cn[, "Average 79T CN"] < rng[2]
gns = gene.names.db[match(gene.cn[idx, "Gene"], gene.ids.db)]
cn.groups.dndscv$`CN=1 (non-essential)` = gns[!is.na(gns) &
                                                  !(gns %in% unlist(essential.genes.dndscv))]

# Run dNdScv
dnds.cn.cv = lapply(1:length(cn.groups.dndscv), function(i) {
    cat("\nGroup: ", names(cn.groups.dndscv)[i], "\n", sep="")
    suppressWarnings(dndscv(mutations, cv=cf3.covs, refdb=INPUT$CF3.CDS,
                            max_muts_per_gene_per_sample=Inf,
                            max_coding_muts_per_sample=Inf, maxcovs=Inf,
                            gene_list=cn.groups.dndscv[[i]], outp=1))
})
names(dnds.cn.cv) = names(cn.groups.dndscv)


## (5) Run dNdScv per abundance quintile group
cat("\nCalculating global dN/dS per transcript abundance quintile...\n")

# Make gene lists
abundance.groups.dndscv = lapply(genes.per.quintile, function(genes) {
    gene.names.db[na.omit(match(genes, gene.ids.db))]
    
})
names(abundance.groups.dndscv) = c("<Q1", "Q1-Q2", "Q2-Q3", "Q3-Q4", ">Q4")

# Run dNdScv
dnds.abundance.cv = lapply(1:length(abundance.groups.dndscv), function(i) {
    cat("\nGroup: ", names(abundance.groups.dndscv)[i], "\n", sep="")
    suppressWarnings(dndscv(mutations, cv=cf3.covs, refdb=INPUT$CF3.CDS,
                            max_muts_per_gene_per_sample=Inf,
                            max_coding_muts_per_sample=Inf, maxcovs=Inf,
                            gene_list=abundance.groups.dndscv[[i]], outp=1))
})
names(dnds.abundance.cv) = names(abundance.groups.dndscv)


## (6) Run dNdScv per GO group
cat("\nCalculating global dN/dS per gene ontology category...\n")

keywords = c("repair", "damage", "transcription", "translation|ribosom", "mitochondri",
             "proteolysis", "metaboli", "exosom", "immun", "adhesion", "chromatin", "inflammat",
             "vesicle", "proliferati", "apopto", "angiogenesis", "chromosom", "folding|folded",
             "cytoskeleton", "cell migration", "cell cycle", "phagosome|phagy")

go.genes.dndscv = sapply(keywords, function(key) {
    gns = gene.names.db[match(unique(cf3.go$`Gene stable ID`[grep(key, cf3.go$`GO term name`)]),
                              gene.ids.db)]
    gns[!is.na(gns)]
})

# Run dNdScv
dnds.go.cv = lapply(1:length(go.genes.dndscv), function(i) {
    cat("\nGroup: \"", names(go.genes.dndscv)[i], "\"\n", sep="")
    suppressWarnings(dndscv(mutations, cv=cf3.covs, refdb=INPUT$CF3.CDS,
                            max_muts_per_gene_per_sample=Inf,
                            max_coding_muts_per_sample=Inf, maxcovs=Inf,
                            gene_list=go.genes.dndscv[[i]], outp=1))
})
names(dnds.go.cv) = names(go.genes.dndscv)

# Get p-values and apply Bonferroni correction
dnds.go.pvals = sapply(dnds.go.cv, function(dndsout) {
    summary(dndsout$poissmodel)$coefficients[c("wmis", "wnon"), "Pr(>|z|)"]
})
dnds.go.pvals.adj = t(apply(dnds.go.pvals, 1, function(row) {
    p.adjust(row, method="bonferroni")
}))


## (7) Plot global dN/dS
cat("\nPlotting global dN/dS figure to output directory...\n")

dnds.list = c(list(dnds.all.cv), dnds.essential.cv[2:3], dnds.cn.cv[1:3], dnds.abundance.cv)

mle = sapply(dnds.list, function(dndsout) {
    dndsout$globaldnds[c("wmis", "wnon"), "mle"]
})

upper = sapply(dnds.list, function(dndsout) {
    dndsout$globaldnds[c("wmis", "wnon"), "cihigh"]
})

lower = sapply(dnds.list, function(dndsout) {
    dndsout$globaldnds[c("wmis", "wnon"), "cilow"]
})

rect.x = c(0, seq(2.5, 21.5, 2), 23)
rect.y = c(0, 1.2)
dot.x = 1:length(mle)
dot.x[dot.x %% 2 == 0] = dot.x[dot.x %% 2 == 0] - 0.4
dot.x[dot.x %% 2 != 0] = dot.x[dot.x %% 2 != 0] + 0.2
cols = c("deepskyblue3", "tomato")
plot.cols =rep(cols, 11)
rect.cols = rep(c("grey90", "grey97"), 6)

pdf(OUTPUT$PLOTS, 12, 4)
par(mar = c(1, 4.5, 2, 1))
plot(dot.x, mle, ylim=c(0.0444, 1.1557), xlim=c(1.3,21.7), las=2, ylab="dN/dS",
     xaxt="n", xlab="", type="n", cex.lab=1.2)
for (i in 1:(length(rect.x)-1)) {
    polygon(x=rect.x[c(i, i+1, i+1, i)], y=rep(rect.y, each=2), border=NA, col=rect.cols[i])
}
abline(v=c(2.5, 6.5, 12.5))
abline(h=1, lty=2)
points(dot.x, mle, pch=16, col=plot.cols, cex=1.5)
box()

arrows(dot.x, mle, dot.x, lower, length=0, lwd=3, col=plot.cols)
arrows(dot.x, mle, dot.x, upper, length=0, lwd=3, col=plot.cols)
abline(v=which(is.na(mle)))
mtext(c("All genes", "Essential genes", "Genes per copy number",
        "Genes per expression quintile category"),
      at=c(1.5, 4.5, 9.5, 17.5), line=0.5, cex=1.2)
mtext(names(dnds.list)[-1], side=1, line=-2.2, at=seq(3.5, 21.5, 2), cex=1.1)
mtext(paste(c(length(RefCDS), lengths(essential.genes.dndscv)[-1],
              lengths(cn.groups.dndscv)[1:3], lengths(abundance.groups.dndscv)), "genes"),
      at=seq(1.5, 21.5, 2), side=1, line=-1.2, cex=0.95)
mtext(c("Missense", "Nonsense"), side=1, at=1.5, line=-5:-4, col=cols, font=2, cex=1.1)
invisible(dev.off())


## (8) Output dN/dS tables
cat("\nWriting dN/dS tables to output directory...\n")
types = c("wmis", "wnon", "wall")

# We omit dN/dS estimates for splice-site mutations because of an intrinsic
# bias of the model when applied to this type of mutations in the dog exome.
# This is likely due to discrepancies between the reference exome regions
# used by the model and the regions that were actually targeted in the study.
cat("GLOBAL DN/DS FOR MISSENSE (wmis), NONSENSE (wnon) AND ALL (wall) SOMATIC SUBSTITUTIONS\n",
    "\nALL GENES (n = ", length(RefCDS), ")\n",
    file=OUTPUT$GLOBAL.TBL, sep="")
suppressWarnings(write.table(cbind("name"=dnds.all.cv$globaldnds[types, 1],
                                   round(dnds.all.cv$globaldnds[types, -1], 5)),
                             file=OUTPUT$GLOBAL.TBL,
                             sep="\t", quote=FALSE, row.names=FALSE, append=TRUE))
cat("\n", file=OUTPUT$GLOBAL.TBL, append=TRUE)

for (i in 1:length(dnds.essential.cv)) {
    cat("\nESSENTIAL GENES (", names(dnds.essential.cv)[i],
        ") (n = ", lengths(essential.genes.dndscv)[i], ")\n",
        file=OUTPUT$GLOBAL.TBL, sep="", append=TRUE)
    suppressWarnings(write.table(cbind("name"=dnds.essential.cv[[i]]$globaldnds[types, 1],
                                       round(dnds.essential.cv[[i]]$globaldnds[types, -1], 5)),
                                 file=OUTPUT$GLOBAL.TBL,
                                 sep="\t", quote=FALSE, row.names=FALSE, append=TRUE))
}
cat("\n", file=OUTPUT$GLOBAL.TBL, append=TRUE)

for (i in 1:length(dnds.cn.cv)) {
    cat("\nPER COPY NUMBER: GENES WITH ", names(dnds.cn.cv)[i],
        " (n = ", lengths(cn.groups.dndscv)[i], ")\n",
        file=OUTPUT$GLOBAL.TBL, sep="", append=TRUE)
    suppressWarnings(write.table(cbind("name"=dnds.cn.cv[[i]]$globaldnds[types, 1],
                                       round(dnds.cn.cv[[i]]$globaldnds[types, -1], 5)),
                                 file=OUTPUT$GLOBAL.TBL,
                                 sep="\t", quote=FALSE, row.names=FALSE, append=TRUE))
}
cat("\n", file=OUTPUT$GLOBAL.TBL, append=TRUE)

for (i in 1:length(dnds.abundance.cv)) {
    cat("\nPER GENE EXPRESSION QUINTILE: GENES WITH ABUNDANCE IN ",
        names(dnds.abundance.cv)[i], " (n = ", lengths(abundance.groups.dndscv)[i], ")\n",
        file=OUTPUT$GLOBAL.TBL, sep="", append=TRUE)
    suppressWarnings(write.table(cbind("name"=dnds.abundance.cv[[i]]$globaldnds[types, 1],
                                       round(dnds.abundance.cv[[i]]$globaldnds[types, -1], 5)),
                                 file=OUTPUT$GLOBAL.TBL,
                                 sep="\t", quote=FALSE, row.names=FALSE, append=TRUE))
}
cat("\n", file=OUTPUT$GLOBAL.TBL, append=TRUE)

for (i in 1:length(dnds.go.cv)) {
    cat("\nPER GENE ONTOLOGY GROUP: GENES WITH GO TERMS CONTAINING \"",
        names(dnds.go.cv)[i], "\" (n = ", lengths(go.genes.dndscv)[i], ")\n",
        file=OUTPUT$GLOBAL.TBL, sep="", append=TRUE)
    suppressWarnings(write.table(cbind("name"=dnds.go.cv[[i]]$globaldnds[types, 1],
                                       round(dnds.go.cv[[i]]$globaldnds[types, -1], 5)),
                                 file=OUTPUT$GLOBAL.TBL,
                                 sep="\t", quote=FALSE, row.names=FALSE, append=TRUE))
}
cat("\n", file=OUTPUT$GLOBAL.TBL, append=TRUE)

cat("\nP-VALUES PER GENE ONTOLOGY GROUP (ADJUSTED VIA BONFERRONI CORRECTION)\n",
    file=OUTPUT$GLOBAL.TBL, sep="", append=TRUE)
suppressWarnings(write.table(cbind("group"=colnames(dnds.go.pvals.adj),
                                   t(round(dnds.go.pvals.adj, 5))),
                             file=OUTPUT$GLOBAL.TBL,
                             sep="\t", quote=FALSE, row.names=FALSE, append=TRUE))

# Output dN/dS per gene
write.table(dnds.all.cv$sel_cv, file=OUTPUT$GENE.TBL,
            sep="\t", quote=FALSE, row.names=FALSE)


cat("\nSaving generated objects to file ", OUTPUT$RDATA, "...\n", sep="")
save(dnds.all.cv, dnds.essential.cv, dnds.abundance.cv, dnds.cn.cv, dnds.go.cv, dnds.go.pvals.adj,
     file=OUTPUT$RDATA)

cat("\nDone\n\n")
