# Analysis scripts for Baez-Ortega et al., 2019
# Step 3. Process variant annotation

# Adrian Baez-Ortega, 2018


# TO RUN THIS SCRIPT IN THE TERMINAL
# ----------------------------------
# Run the commands below in the terminal, replacing '/path/to/abo2018' with 
# the path to the abo2018 directory.
#
#    cd /path/to/abo2018
#    Rscript scripts/3_ProcessAnnotation.R


# TO RUN THIS SCRIPT IN RSTUDIO
# -----------------------------
# Before starting, run the line below in RStudio, replacing '/path/to/abo2018' with 
# the path to the abo2018 directory.
#
#    setwd("path/to/abo2018")


# If the paths to the input or output files differ from the ones 
# used in this script, they may be updated by modifying the lines below.

INPUT = list(
    
    # Paths to input files containing variant annotation for gemline and somatic SNVs/indels
    # (from Ensembl's Variant Effect Predictor)
    VEP.SNV.GL = file.path("data", "original", "VEP_Annotation_SNVs_Germline.tsv"),
    VEP.SNV.SM = file.path("data", "original", "VEP_Annotation_SNVs_Somatic.tsv"),
    VEP.IND.GL = file.path("data", "original", "VEP_Annotation_Indels_Germline.tsv"),
    VEP.IND.SM = file.path("data", "original", "VEP_Annotation_Indels_Somatic.tsv"),
    
    # Paths to input files containing copy number segments for samples 24T and 79T
    # (from Murchison et al., Science 2014, in CanFam2 coordinates)
    CN.24T = file.path("data", "original", "Copy_Number_CTVT_24T.tsv"),
    CN.79T = file.path("data", "original", "Copy_Number_CTVT_79T.tsv"),
    
    # Path to input file containing SNV position correspondence between CanFam2 and CanFam3.1
    # (from NCBI's Genome Remapping Service)
    CF2.POS = file.path("data", "original", "CanFam2_Positions_SNVs.tsv"),
    
    # Path to input file containing ID correspondence between CanFam3.1 genes and transcripts
    CF3.TX = file.path("data", "original", "CanFam3.1_Transcripts.tsv"),
    
    # Path to input file generated by script 2_ImportVariants.R
    VAR.TABLES = file.path("data", "processed", "Variant_Tables.RData")
    
)

OUTPUT = list(
    
    # Path to output RData file
    RDATA = file.path("data", "processed", "Variant_Annotation.RData")
    
)


# Print input and output file paths
cat("\nInput files:")
for (name in INPUT) {
    cat("\n  ", name)
}
cat("\n\nOutput files:")
for (name in OUTPUT) {
    cat("\n  ", name)
}
cat("\n\n")


# Load packages
library(stringr)


# Load input data
cat("Loading data...\n")

# We refer to somatic and germline variants as 'tumour-only' (or 'tonly') and 'host+tumour' (or 'host'), resp.
annot.snvs.tonly.orig = read.table(INPUT$VEP.SNV.SM, 
                                   skip=20, comment.char="", header=T, stringsAsFactors=F)    # Somatic SNVs
annot.snvs.host.orig = read.table(INPUT$VEP.SNV.GL, 
                                  skip=20, comment.char="", header=T, stringsAsFactors=F)     # Germline SNVs
annot.indels.tonly.orig = read.table(INPUT$VEP.IND.SM, 
                                     skip=20, comment.char="", header=T, stringsAsFactors=F)  # Somatic indels
annot.indels.host.orig = read.table(INPUT$VEP.IND.GL, 
                                    skip=20, comment.char="", header=T, stringsAsFactors=F)   # Germline indels
colnames(annot.snvs.tonly.orig)[1] = 
    colnames(annot.snvs.host.orig)[1] = 
    colnames(annot.indels.tonly.orig)[1] = 
    colnames(annot.indels.host.orig)[1] = "Uploaded_variation"

cn.24T = read.table(INPUT$CN.24T, header=T, stringsAsFactors=F)
cn.79T = read.table(INPUT$CN.79T, header=T, stringsAsFactors=F)
cf2.pos.corresp = read.table(INPUT$CF2.POS, header=T, fill=T, stringsAsFactors=F)
transcripts = read.table(INPUT$CF3.TX, header=T, stringsAsFactors=F)
load(INPUT$VAR.TABLES)


# Using the available copy number data for 24T and 79T (in CanFam2 coordinates), and the 
# position correspondence table between CanFam2 and CanFam3.1, build a table with all SNVs
# and the fields: CanFam3.1 chrom, CanFam3.1 position, CanFam2 chrom, CanFam2 position, 
# integer total CN (24T), integer total CN (79T).
cat("Assigning copy number estimates to all SNVs...\n")

cn.table.snvs = as.data.frame(t(
    sapply(1:nrow(cf2.pos.corresp), function(i) {
    
        var = cf2.pos.corresp[i, ]
        cn24 = cn79 = NA
    
        # If the variant has no CanFam2 coords, leave CN=NA
        if (var$CF2.CHR == "NULL") {
            var$CF2.POS = "NULL"
        }
        
        # Otherwise, retrieve integer total copy number in 24T and 79T
        else {
            idx = cn.24T$CHR == var$CF2.CHR & 
                cn.24T$START <= var$CF2.POS & 
                cn.24T$END >= var$CF2.POS
            if (sum(idx) == 1) {
                cn24 = cn.24T$INT.TOTAL.CN[idx]
            }
            idx = cn.79T$CHR == var$CF2.CHR & 
                cn.79T$START <= var$CF2.POS & 
                cn.79T$END >= var$CF2.POS
            if (sum(idx) == 1) {
                cn79 = cn.79T$INT.TOTAL.CN[idx]
            }
        }
        
        unlist(c(var, "CN.24T"=cn24, "CN.79T"=cn79))
    })
), stringsAsFactors=F)

cn.table.snvs$CF3.POS = as.integer(cn.table.snvs$CF3.POS)
cn.table.snvs$CF2.POS = suppressWarnings(as.integer(cn.table.snvs$CF2.POS))
cn.table.snvs$CN.24T = as.integer(cn.table.snvs$CN.24T)
cn.table.snvs$CN.79T = as.integer(cn.table.snvs$CN.79T)

# Because of a normalisation issue due to the matched host 79H being male, some chrX regions appear
# as having CN=2 in 79T. To account for this, for chrX variants having CN=1 in 24T and CN=2 in 79T, 
# we also set the copy number in 79T to CN=1.
idx = (cn.table.snvs$CF3.CHR == "X" & cn.table.snvs$CN.24T == 1 & cn.table.snvs$CN.79T == 2) %in% TRUE
cn.table.snvs$CN.79T[idx] = 1


# Function to process annotation table
process.annotation = function(var.annot, var.metadata, var.nv, snvs = FALSE, tonly = FALSE) {
    
    # We consider only variants that have not been considered before (non-duplicated) and are in 
    # transcripts that are included in the `transcripts` matrix (containing one transcript per gene)
    included.variants = rep(F, nrow(var.annot))
    for (i in 1:nrow(var.annot)) {
        if (!(var.annot$Uploaded_variation[i] %in% 
              var.annot$Uploaded_variation[included.variants]) &
            var.annot$Feature[i] %in% transcripts$Name) {
            
            included.variants[i] = TRUE
            
        }
    }
    annotation = var.annot[included.variants, ]
    
    # Expand Extra field
    annotation$Impact = NA
    annotation$Strand = NA
    annotation$Symbol = NA
    annotation$Symbol_source = NA
    annotation$HGNC_ID = NA
    annotation$SIFT = NA
    annotation$Exon = NA
    
    for (i in 1:nrow(annotation)) {
        extra = str_split(annotation$Extra[i], ";")[[1]]
        for (field in extra) {
            cat = str_split_fixed(field, "=", 2)[1]
            val = str_split_fixed(field, "=", 2)[2]
            if (cat == "IMPACT") {
                annotation$Impact[i] = val
            } else if (cat == "STRAND") {
                annotation$Strand[i] = val
            } else if (cat == "SYMBOL") {
                annotation$Symbol[i] = val
            } else if (cat == "SYMBOL_SOURCE") {
                annotation$Symbol_source[i] = val
            } else if (cat == "HGNC_ID") {
                annotation$HGNC_ID[i] = val
            } else if (cat == "SIFT") {
                annotation$SIFT[i] = val
            } else if (cat == "EXON") {
                annotation$Exon[i] = val
            }
        }
    }
    annotation$Extra = NULL
    
    # Add Chrom and Pos fields
    annotation$Chrom = str_split_fixed(annotation$Location, ":", 2)[, 1]
    annotation$Pos = as.integer(
        str_split_fixed(
            str_split_fixed(annotation$Location, ":", 2)[, 2], 
            "-", 2)[, 1])
    # Correct deletion positions by 1 bp (for consistency with indels.metadata)
    idx = annotation$Allele == "-"
    annotation$Pos[idx] = annotation$Pos[idx] - 1
    
    # For SNVs, add Ref, CDS_Ref and CDS_Alt fields
    if (snvs) {
        annotation$Ref = str_split_fixed(str_split_fixed(annotation$Uploaded_variation,
                                                         "_", 3)[, 3],
                                         "/", 2)[, 1]
        annotation$CDS_Ref = NA
        annotation$CDS_Alt = NA
        for (i in 1:nrow(annotation)) {
            if (annotation$Codons[i] != "-") {
                j = grep("[[:upper:]]", str_split_fixed(annotation$Codons[i], "", 7)[1:3])
                annotation$CDS_Ref[i] = substr(annotation$Codons[i], j, j)
                annotation$CDS_Alt[i] = substr(annotation$Codons[i], j + 4, j + 4)
            }
        }
    }
    
    # Add meta-index (variant row number in metadata/nr/nv/vaf tables)
    ann.str = paste0(annotation$Chrom, ":", annotation$Pos)
    met.str = paste0(var.metadata$CHROM, ":", var.metadata$POS)
    annotation$Meta_index = match(ann.str, met.str)
    stopifnot(identical(ann.str, met.str[annotation$Meta_index]))
    
    # For SNVs, add copy number estimates
    if (snvs) {
        annotation$CN_24T = cn.table.snvs$CN.24T[annotation$Meta_index]
        annotation$CN_79T = cn.table.snvs$CN.79T[annotation$Meta_index]
    }
    
    # For tumour-only variants, add number of tumours
    if (tonly) {
        annotation$Number_tumours = rowSums(var.nv[annotation$Meta_index, tumours] >= MIN.READS)
    }
    
    # For SNVs, add trinucleotide context
    if (snvs) {
        annotation$Ref_trinucleotide = substr(str_split_fixed(var.metadata$INFO[annotation$Meta_index],
                                                              ";", 20)[, 12],
                                              13, 15)
    }
    
    annotation
}


# Process T-only indels
cat("Processing annotation for tumour-only (somatic) indels...\n")
annot.indels.tonly = process.annotation(annot.indels.tonly.orig, indels.metadata, indels.nv, tonly=T)

# Process host indels
cat("Processing annotation for host (germline) indels...\n")
annot.indels.host = process.annotation(annot.indels.host.orig, indels.metadata, indels.nv)

# Process T-only SNVs
cat("Processing annotation for tumour-only SNVs...\n")
annot.snvs.tonly = process.annotation(annot.snvs.tonly.orig, snvs.metadata, snvs.nv, snvs=T, tonly=T)

# Process host indels
cat("Processing annotation for host SNVs...\n")
annot.snvs.host = process.annotation(annot.snvs.host.orig, snvs.metadata, snvs.nv, snvs=T)


# Save objects
cat("\nSaving generated objects to file ", OUTPUT$RDATA, "...\n", sep="")
save(annot.snvs.tonly, annot.snvs.host, annot.indels.tonly, annot.indels.host,
     file=OUTPUT$RDATA)

cat("\nDone\n\n")

