# Analysis scripts for Baez-Ortega et al., 2019
# Step 4. Perform exploratory variant analyses

# Adrian Baez-Ortega, 2018


# TO RUN THIS SCRIPT IN THE TERMINAL
# ----------------------------------
# Run the commands below in the terminal, replacing '/path/to/abo2018' with 
# the path to the abo2018 directory.
#
#    cd /path/to/abo2018
#    Rscript scripts/4_ExploratoryAnalyses.R


# TO RUN THIS SCRIPT IN RSTUDIO
# -----------------------------
# Before starting, run the line below in RStudio, replacing '/path/to/abo2018' with 
# the path to the abo2018 directory.
#
#    setwd("path/to/abo2018")


# If the paths to the input or output files differ from the ones 
# used in this script, they may be updated by modifying the lines below.

INPUT = list(
    
    # Path to input file containing fraction of bp sequenced with depth >30X per sample
    COV.FRACTION = file.path("data", "original", "30x_Coverage_Fraction.tsv"),
    
    # Path to input file generated by script 2_ImportVariants.R
    VAR.TABLES = file.path("data", "processed", "Variant_Tables.RData")
    
)

OUTPUT = list(
    
    # Paths to output coverage and VAF histogram files
    COV.HIST.T = file.path("output", "Coverage_Histograms_Tumours.pdf"),
    COV.HIST.H = file.path("output", "Coverage_Histograms_Hosts.pdf"),
    VAF.HIST.T = file.path("output", "VAF_Histograms_Tumours.pdf"),
    VAF.HIST.H = file.path("output", "VAF_Histograms_Hosts.pdf"),
    
    # Paths to output QC/variant summary tables
    QC.SUM.H = file.path("output", "QC_Summary_Hosts.tsv"),
    QC.SUM.T = file.path("output", "QC_Summary_Tumours.tsv"),
    VAR.SUM.T = file.path("output", "Variant_Categories_Summary_Tumours.tsv"),
    
    # Path to output RData file
    RDATA = file.path("data", "processed", "Coverage_Purity_QCindex.RData")
    
)


# Print input and output file paths
cat("\nInput files:")
for (name in INPUT) {
    cat("\n  ", name)
}
cat("\n\nOutput files:")
for (name in OUTPUT) {
    cat("\n  ", name)
}
cat("\n\n")


# Load packages
library(stringr)


# Load input data
cat("Loading data...\n")
load(INPUT$VAR.TABLES)
coverage = read.table(INPUT$COV.FRACTION, header=T, stringsAsFactors=F, check.names=F)
coverage = coverage$`31+_cov_fraction`[match(samples, coverage$Sample)]


# Function to plot VAF/coverage histograms
plot.histogram = function(filename, index, snvs.table, max.x, max.y, keyword, xlab) {
    cairo_pdf(filename, 10, 8, onefile=T)
    par(mar=c(5.1, 4.8, 4.1, 0.8))
    
    for (i in which(index)) {
        present = snvs.nv[, i] >= MIN.READS
        h = hist(snvs.table[present, i][snvs.table[present, i] <= max.x],
                 breaks=seq(0, max.x, len=100),
                 col="dodgerblue4", border="white", 
                 xlim=c(0, max.x), ylim=c(0, max.y), xlab=xlab, ylab="SNVs",
                 main=paste0(keyword, " histogram for ", samples[i], "\n",
                             sum(present), " SNVs (≥", MIN.READS, " supporting reads)"))
        if (keyword == "VAF") {
            segments(x0=0.253, y0=0.13, y1=max.y, col="darkgrey", lwd=1.5)
            text(x=0.106, y=0.98 * max.y, labels="Tumour contamination", col="grey40")
        }
    }
    invisible(dev.off())
}


# Plot coverage histograms per host and per tumour (only SNVs)
cat("Plotting coverage histograms to output directory...\n")
plot.histogram(OUTPUT$COV.HIST.H, hosts, snvs.nr, 300, 7000, "Coverage", "Total read coverage")
plot.histogram(OUTPUT$COV.HIST.T, tumours, snvs.nr, 300, 7000, "Coverage", "Total read coverage")


# Plot VAF histograms of germline variants per host (only SNVs)
cat("Plotting VAF histograms to output directory...\n")
plot.histogram(OUTPUT$VAF.HIST.H, hosts, snvs.vaf, 1, 4000, "VAF", "VAF in host")


# Plot VAF histograms of categorised variants per tumour (only SNVs)
# For each tumour, plot (on the same page) the VAF of variants that are:
#  - Tumour-only
#  - Homozygous in matched host (VAF>0.75 in matched host)
#  - Heterozygous in matched host (0.25<VAF<0.75 in matched host)
#  - Contamination in matched host (VAF<0.25 in matched host)
#  - Not in matched host (i.e. variant does not fit any of the categories above)

# Take into account the special case of 1224T2, whose matched host is not 
# located inmediately before in the variant tables
special = data.frame("tumour.idx" = match("1224T2", samples),
                     "host.idx" = match("1224H", samples))

# Take into account tumours without matched hosts
tumour.num = str_split_fixed(samples[tumours], "T", 2)[, 1]
host.num = str_split_fixed(samples[hosts], "H", 2)[, 1]
unmatched = rep(FALSE, length(samples))
unmatched[tumours][!(tumour.num %in% host.num)] = TRUE
unmatched = which(unmatched)

# Plot histograms and collect variant numbers
cairo_pdf(OUTPUT$VAF.HIST.T, 20, 5, onefile=T)
par(mfrow=c(1, 5), mar=c(5.1, 2.1, 4.1, 1.1), oma=c(0, 3, 4, 1))
MAX.Y = 4000

variant.categ = t(sapply(which(tumours), function(i) {
    
    # Get index of matched host
    matched.host = i - 1
    # Take into account special samples 
    if (i %in% special$tumour.idx) {
        matched.host = special$host.idx[special$tumour.idx == i]
    }
    
    # Get present variants
    present = snvs.nv[, i] >= MIN.READS
    
    # Plot variants in each category:
    # T-only
    present.Tonly = present & tumour.only.idx
    hist(snvs.vaf[present.Tonly, i], 
         breaks=seq(0, 1, 0.01), col="dodgerblue4", border="white", xpd=NA,
         xlab="VAF in tumour", ylab="SNVs", ylim=c(0, MAX.Y), xlim=c(0, 1),
         main=paste0(sum(present.Tonly), " tumour-only SNVs (≥", 
                     MIN.READS, " supporting reads)"))
    
    # If no matched host: plot nothing
    if (i %in% unmatched & !(i %in% special$tumour.idx)) {
        host.avail = "N"
        host.indices = matrix(NA, nrow=1, ncol=4)
        for (j in 1:4) {
            hist(1, border=NA, xlab="VAF in tumour", ylab="", 
                 ylim=c(0, MAX.Y), xlim=c(0, 1), main="(No matched host)")
        }
    }
    
    # If matched host available: plot non-tumour-only variants
    else {
        host.avail = "Y"
        host.indices = cbind(
            # Found in tumour & homozygous in matched host
            "homozygous" = present & snvs.vaf[, matched.host] > 0.75,
            
            # Found in tumour & heterozygous in matched host
            "heterozygous" = present & snvs.vaf[, matched.host] > 0.25 & snvs.vaf[, matched.host] < 0.75,
            
            # Found in tumour & contamination in matched host
            "contamination" = present & snvs.vaf[, matched.host] < 0.25 & snvs.vaf[, matched.host] > 0,
            
            # Found in tumour & not in matched host
            "not present" = present & !tumour.only.idx & snvs.vaf[,matched.host] == 0
        )
        
        for (j in 1:ncol(host.indices)) {
            hist(snvs.vaf[host.indices[, j], i], 
                 breaks=seq(0, 1, 0.01), col="dodgerblue4", border="white", 
                 xlab="VAF in tumour", ylab="", ylim=c(0, MAX.Y), xlim=c(0, 1),
                 main=paste0(sum(host.indices[, j]), " SNVs (≥", 
                             MIN.READS, " supporting reads) that are\n", 
                             colnames(host.indices)[j], " in ", samples[matched.host]))
        }
    }
    
    title(paste("VAF histograms for", samples[i]), outer=T, cex=1.2)
    
    c(samples[i], sum(present.Tonly), colSums(host.indices), host.avail)
}))

invisible(dev.off())


# Output table of variants per tumour and category
cat("Writing tumour variant summary to output directory...\n")
colnames(variant.categ) = c("Sample", "T-only SNVs", 
                            "SNVs present in T and homozygous in H", 
                            "SNVs present in T and heterozygous in H",
                            "SNVs present in T and contamination in H",
                            "SNVs present in T and not in H",
                            "Matched H available")
write.table(variant.categ, row.names=F, quote=F, sep="\t",
            file=OUTPUT$VAR.SUM.T)


# Obtain tumour purity estimates (from T-only VAF histogram modes)
# For each tumour, obtain the VAF value corresponding to the mode of 
# the tumour VAF histogram for present T-only variants
purity = sapply(which(tumours), function(i) {
    present = tumour.only.idx & 
        snvs.nv[, i] >= MIN.READS
    hist.data = hist(snvs.vaf[present, i], breaks=seq(0, 1, 0.01), plot=F)
    2 * hist.data$mids[which.max(hist.data$counts)]
})


# Obtain tumour 'QC index', defined as the product (coverage * purity)
qc.index = purity * coverage[tumours]


# Output QC summaries for hosts and tumours
cat("Writing sample QC summaries to output directory...\n")

hosts.sum = cbind(samples[hosts],
                  t(sapply(which(hosts), function(i) {
                        c(sum(snvs.nv[, i] >= MIN.READS),
                          sum(indels.nv[, i] >= MIN.READS))
                    })),
                  coverage[hosts])
colnames(hosts.sum) = c("Sample", "SNVs", "Indels", "31+ coverage fraction")
write.table(hosts.sum, quote=F, row.names=F, sep="\t",
            file=OUTPUT$QC.SUM.H)

tumours.sum = cbind(samples[tumours],
                    t(sapply(which(tumours), function(i) {
                        c(sum(snvs.nv[tumour.only.idx, i] >= MIN.READS),
                          sum(tumour.unique.idx[, samples[i]]),
                          sum(indels.nv[tumour.only.indels.idx, i] >= MIN.READS),
                          sum(tumour.unique.indels.idx[, samples[i]]))
                      })),
                    coverage[tumours], 
                    purity,
                    qc.index)
colnames(tumours.sum) = c("Sample", "Tumour-only SNVs", "Tumour-unique SNVs", 
                          "Tumour-only indels", "Tumour-unique indels",
                          "31+ coverage fraction", "Purity", "QC index")
write.table(tumours.sum, quote=F, row.names=F, sep="\t",
            file=OUTPUT$QC.SUM.T)


# Save objects
cat("\nSaving generated objects to file ", OUTPUT$RDATA, "...\n", sep="")
save(coverage, purity, qc.index,
     file=OUTPUT$RDATA)

cat("\nDone\n\n")

