# Analysis scripts for Baez-Ortega et al., 2019
# Step 7. Mutational signature analyses

# Adrian Baez-Ortega, 2018


# TO RUN THIS SCRIPT IN THE TERMINAL
# ----------------------------------
# Run the commands below in the terminal, replacing '/path/to/TCG2019' with
# the path to the TCG2019 directory.
#
#    cd /path/to/TCG2019
#    Rscript scripts/7_SignatureAnalyses.R


# TO RUN THIS SCRIPT IN RSTUDIO
# -----------------------------
# Before starting, run the line below in RStudio, replacing '/path/to/TCG2019' with
# the path to the TCG2019 directory.
#
#    setwd("path/to/TCG2019")


# If the paths to the input or output files differ from the ones
# used in this script, they may be updated by modifying the lines below.

INPUT = list(
    
    # Path to input file generated by script 2_ImportVariants.R
    VAR.TABLES = file.path("data", "processed", "Variant_Tables.RData"),
    
    # Path to input file generated by script 5_PhyloGroups.R
    PHYLO.GROUPS = file.path("data", "processed", "Phylo_Tree_Groups.RData")
    
)

OUTPUT = list(
    
    # Path to output directory for plots
    PLOTS.DIR = file.path("output", "Mutational_Signatures_Plots"),
    
    # Path to output signatures table
    SIGNATURES = file.path("output", "Mutational_Signatures_Table.tsv"),
    
    # Path to output RData file
    RDATA = file.path("data", "processed", "Signatures_Exposures.RData")
    
)


# Print input and output file paths
cat("\nInput files:")
for (name in INPUT) {
    cat("\n  ", name)
}
cat("\n\nOutput files:")
for (name in OUTPUT) {
    cat("\n  ", name)
}
cat("\n\n")


# Load packages
library(stringr)
library(sigfit, quietly=T)


# Load input data
cat("Loading data...\n")
load(INPUT$VAR.TABLES)
load(INPUT$PHYLO.GROUPS)


# Generate mutational spectrum of each phylogenetic group
cat("Generating mutational spectra of phylogenetic groups...\n")

# Build table containing group ID, ref base, alt base, and
# trinucleotide context for all group-unique variants
trinuc = substr(str_split_fixed(snvs.metadata$INFO, ";", 20)[, 12], 13, 15)
var.table = matrix(NA, nrow=sum(group.unique.idx), ncol=4,
                   dimnames=list(NULL, c("ID", "REF", "ALT", "CONTEXT")))
i = 1
for (j in 1:ncol(group.unique.idx)) {
    k = i + sum(group.unique.idx[, j]) - 1
    var.table[i:k, ] = cbind(colnames(group.unique.idx)[j],
                             snvs.metadata$REF[tumour.only.idx][group.unique.idx[, j]],
                             snvs.metadata$ALT[tumour.only.idx][group.unique.idx[, j]],
                             trinuc[tumour.only.idx][group.unique.idx[, j]])
    i = k + 1
}

# Build spectra (sigfit)
group.spectra = sigfit::build_catalogues(var.table)


## NB. The commented block of code below performs preliminary extraction of
## N=3 signatures from the set of all spectra with >700 mutations, and plots 
## the results. As these results were found to be suboptimal for some groups 
## (notably A1), this step is ommited by default.

# # Extract mutational signatures from all groups with >700 mutations
# # (i.e. all groups except 26, 39, 47, 55)
# MIN.MUT = 700
# MAX.MUT = 2000
# exclude.idx = rowSums(group.spectra) < MIN.MUT
# counts = group.spectra[!exclude.idx, ]
# # Downsample to a maximum of 2000 mutations
# for (i in which(rowSums(counts) > MAX.MUT)) {
#     counts[i, ] = round(counts[i, ] / sum(counts[i, ]) * MAX.MUT)
# }
# # Extract 2-5 signatures
# N = 2:5
# cat("\nPreliminary sigfit run: extracting ", N[1], "-", tail(N, 1),
#     " signatures from all groups with >", MIN.MUT, " mutations\n", sep="")
# cat("Excluded groups:", paste(which(exclude.idx), collapse=", "), "\n\n")
# sigfit.prelim = sigfit::extract_signatures(counts, nsignatures=N,
#                                            iter=1000, seed=1756)
# cat("\nPlotting preliminary (suboptimal) results for extraction of N=3 signatures to output directory...\n")
# invisible(sigfit::plot_all(sigfit.prelim[[3]],
#                            out_path=file.path("output", "Preliminary_Signature_Extraction")))


# Extract N=3 signatures from non-ancestral groups with >700 mutations
# (i.e. all groups except 26, 39, 47, 55, A1, A2, A3), downsampling to
# a maximum of 2000 mutations
# (The extracted signatures correspond to COSMIC signatures 1, 5 and 7)
MIN.MUT = 700
MAX.MUT = 2000
exclude.idx = (rowSums(group.spectra) < MIN.MUT) | (group.ids %in% c("A1", "A2", "A3"))
counts = group.spectra[!exclude.idx, ]
for (i in which(rowSums(counts) > MAX.MUT)) {
    counts[i, ] = round(counts[i, ] / sum(counts[i, ]) * MAX.MUT)
}

# Extract 3 signatures
N = 3
cat("\nExtracting ", N, " signatures from non-ancestral groups with >",
    MIN.MUT, " mutations\n", sep="")
cat("Excluded groups:", paste(group.ids[exclude.idx], collapse=", "), "\n\n")

sigfit.extract.3 = sigfit::extract_signatures(counts, nsignatures=N,
                                              iter=10000, warmup=3000, seed=1756,
                                              control=list(adapt_delta=0.85))

# Retrieve and reorder extracted signatures
signatures.1.5.7 = sigfit::retrieve_pars(sigfit.extract.3, "signatures")$mean[3:1, ]


# Extract the signatures accounting for the additional patterns in groups
# 57, 58, A1, A3, by using sigfit to simultaneously fit the present three
# signatures and extract two additional ones
# Use only groups 57, 58, A1, A3, downsampling to a maximum of 5000 mutations
# (The extracted signatures correspond to signatures A and 2* in the paper)
MAX.MUT = 5000
select.idx = match(c("57", "58", "A1", "A3"), group.ids)
counts = group.spectra[select.idx, ]
for (i in which(rowSums(counts) > MAX.MUT)) {
    counts[i, ] = round(counts[i, ] / sum(counts[i, ]) * MAX.MUT)
}

# Fit 3 signatures, extract 2 signatures
N = 2
cat("\nExtracting", N, "additional signatures from groups:",
    paste(group.ids[select.idx], collapse=", "), "\n\n")

# (Using Jeffreys prior on signatures and exposures)
sigfit.fitext.2 = sigfit::fit_extract_signatures(counts, signatures=signatures.1.5.7,
                                                 num_extra_sigs=N, exp_prior=0.5,
                                                 sig_prior=matrix(0.5, nrow=N, ncol=96),
                                                 iter=10000, warmup=3000, seed=1756)

# Retrieve and rename extracted signatures
signatures.final = sigfit::retrieve_pars(sigfit.fitext.2, "signatures")$mean
rownames(signatures.final) = paste("Signature", c("1", "5", "7", "2*", "A"))


## NB. The commented block of code below performs preliminary re-fitting of all
## extracted signatures to all phylogenetic groups to obtain preliminary signature 
## exposures, and plots the results. As these results show that signatures 2* and 
## A are not active in most groups, this step is ommited by default.

# # Fit mutational signatures to all groups
# sigfit.fit.prelim = sigfit::fit_signatures(group.spectra, signatures.final,
#                                            iter=4000, warmup=2000, chains=2, seed=1756)
# cat("\nPlotting preliminary (suboptimal) results for refitting of all signatures to output directory...\n")
# invisible(sigfit::plot_all(sigfit.fit.prelim, exp_legend_pos="topleft",
#                            out_path=file.path("output", "Preliminary_Signature_Fitting")))


# Re-fit extracted signatures to obtain refined signature exposures
# Based on a previous re-fitting of all 5 signatures to all phylogenetic groups
# (see above), we determined the sets of signatures with significant activity 
# (defined as lower bound >0.01 for the HPD interval of signature exposure)
# in each group:
#  - Signatures 1, 5, 7 are active in all groups
#  - Signature 2* is active in groups 57, 58, A3
#  - Signature A is active in groups A1, A2, A3

refine.exposures = function(sigs.idx, groups.idx) {
    cat("\nRefitting", paste(rownames(signatures.final)[sigs.idx], collapse=", "),
        "to groups:", paste(group.ids[groups.idx], collapse=", "), "\n\n")
    
    sigfit.fit = sigfit::fit_signatures(group.spectra[groups.idx, ],
                                        signatures.final[sigs.idx, ],
                                        iter=4000, warmup=2000, seed=1756,
                                        control=list(adapt_delta=0.85))
    sigfit::retrieve_pars(sigfit.fit, "exposures")
}

# Re-fit signatures {1, 5, 7} to all groups except {57, 58, A1, A2, A3}
exposures.general = refine.exposures(1:3, !(group.ids %in% c("57", "58", "A1", "A2", "A3")))
    
# Re-fit signatures {1, 5, 7, 2*} to groups {57, 58}
exposures.57.58 = refine.exposures(1:4, group.ids %in% c("57", "58"))

# Re-fit signatures {1, 5, 7, A} to groups {A1, A2}
exposures.A1.A2 = refine.exposures(-4, group.ids %in% c("A1", "A2"))

# Re-fit signatures {1, 5, 7, 2*, A} to group A3
exposures.A3 = refine.exposures(1:5, group.ids == "A3")


# Make definitive exposures table
# (NB. The information in this table will be output in Step 9)
exposures.final = as.data.frame(
    matrix(NA, nrow=nrow(group.spectra), ncol=nrow(signatures.final) * 3,
           dimnames=list(group.names,
                         sapply(c("(mean)", "(lower 95%)", "(upper 95%)"), function(x) {
                             paste(rownames(signatures.final), "exposure", x)
                         }))))

fetch.exposure = function(exposures, name) {
    sapply(exposures, function(e) {
        if (name %in% colnames(e)) {
            return(e[, name])
        }
        integer(nrow(e))
    })
}

for (name in rownames(signatures.final)) {
    idx = grep(name, colnames(exposures.final), fixed=T)
    exposures.final[, idx] = rbind(fetch.exposure(exposures.general, name),
                                   fetch.exposure(exposures.57.58, name),
                                   fetch.exposure(exposures.A1.A2, name),
                                   fetch.exposure(exposures.A3, name))
}


# Plot signatures and exposures
cat("\nWriting definitive results of mutational signature analysis to output directory...\n")
cat("(Table of signature exposures will be written in Step 9)\n\n")
exposures.plot = list(mean=exposures.final[, grep("mean", colnames(exposures.final))],
                      lower_95=exposures.final[, grep("lower", colnames(exposures.final))],
                      upper_95=exposures.final[, grep("upper", colnames(exposures.final))])
for (i in 1:length(exposures.plot)) {
    colnames(exposures.plot[[i]]) = rownames(signatures.final)
}

invisible(sigfit::plot_all(counts=group.spectra, signatures=signatures.final, 
                           exposures=exposures.plot, out_path=OUTPUT$PLOTS.DIR, prefix="Definitive",
                           rec_legend_pos="topleft", exp_legend_pos="topleft", exp_margin_bottom=17.5))

# Output signatures table
write.table(cbind("Mutation type"=colnames(signatures.final), t(signatures.final)),
            file=OUTPUT$SIGNATURES, sep="\t", quote=F, row.names=F)


cat("\nSaving generated objects to file ", OUTPUT$RDATA, "...\n", sep="")
save(group.spectra, signatures.final, exposures.final, file=OUTPUT$RDATA)

cat("\nDone\n\n")
